### PR TITLE
Tracker v2.1 - Harder, Better, Faster, Stronger, Threadsafe

### PR DESCRIPTION
**Pull request by d4rken**
(I somehow broke the previous pullrequest)

The base goal was to make the Tracker class thread safe.

Previously the tracker was the sole responsible entity for the queryparamters.
Multiple threads using the same tracker would overwrite each others values.
Also multiple trackers were no solution as each tracker would automatically start a new session.

The idea is to decouple the *Tracker* from holding specific query data.
It should know where to deliver the queries and supplement basic information if necessary, but not more.
Each thread/request/actor constructs it's own query then passes it to the Tracker.
The Tracker will inspect the query, fill in missing information and then give it to the *Dispatcher* to be dispatched.

Change notes:

* A new tracking session starts after the first track, not on *Tracker* creation. Because we can always reuse the same *Tracker*, all events stay in the same visitor session.
* The base method for all tracking is *getTracker().track(TrackMe)* it takes a *TrackMe* object.
* The *TrackMe* object is a wrapper for the *QueryParams* and the screen scope custom user variables. Google Analytics has a similar *HitBuilder* object.
* All other methods like trackScreen/trackEvent just create a *TrackMe* object and set the correct parameters. I.e. they are convinience methods.
* The *track(TrackMe)* method will insert basic information such as screen resolution and useragent only on the first track of the visit (if not already set). While a dev can force set these values, they normally don't change and Piwik just needs them to be transmitted once per visit session.
* The *track(TrackMe)* method will for each visit insert the basic parameters such visitorId/userId etc. and visitor scope variables.
* The *track(TrackMe)* method will then call "build()" on the *TrackMe* object which creates a String with the query.
* This query is then given to the *Dispatcher*, which is a threadsafe queue with a single worker thread.
* The *Dispatcher* will at each interval drain the queue, transmit the events, then go to sleep again.

* Having to set the screen custom variables before tracking a screen is no longer a possible mistake, because you have to create and set them on a *TrackMe* object before you can pass it.

* All time values are now in milliseconds, because everything is always in milliseconds in Android. Having to check wether it is seconds or milliseconds just creates more errors. Dealing with developers not users, milliseconds are perfectly fine.

* There exists a default *TrackMe* object now which is populated on Tracker creation with all necessary parameters. Before each TrackMe object gets send, the Tracker will try to fill in missing parameters from the default TrackMe, but it won't overwrite any existing values. An advanced developer can still use *getDefaultTrackMe* and modify the default values to achieve anykind of customization.

* Methods such as *setResolution* or *setUserAgent* etc. have been removed as the developer is not encouraged to alter them. But they can still be changed through *getDefaultTrackMe*

* A few freedoms are given back to developers now. If a value gets explicitly NULLed in the default TrackMe object, the Tracker will no longer refill it.

* The tests have been adapted to all changes

* A new testclass *TestDispatcher* now exists that does multithreading & concurrency tests.

General Notes:
* Because the *Dispatcher* is now the sole exit point for all queries to be send, it would be a good target to look at for implementing offline mode (#16).

````

D/launchTestThreads: Launching 100 threads, 1000 queries each
D/launchTestThreads: All launched.
D/checkForMIAs: 5216 events created, 331 requests dispatched, containing 4821 flattened queries
D/checkForMIAs: 10923 events created, 705 requests dispatched, containing 10474 flattened queries
D/checkForMIAs: 16695 events created, 1081 requests dispatched, containing 16116 flattened queries
D/checkForMIAs: 22372 events created, 1449 requests dispatched, containing 21975 flattened queries
D/checkForMIAs: 28038 events created, 1817 requests dispatched, containing 27605 flattened queries
D/checkForMIAs: 33916 events created, 2201 requests dispatched, containing 33287 flattened queries
D/checkForMIAs: 39640 events created, 2571 requests dispatched, containing 39158 flattened queries
D/checkForMIAs: 45351 events created, 2945 requests dispatched, containing 44889 flattened queries
D/checkForMIAs: 51173 events created, 3317 requests dispatched, containing 50605 flattened queries
D/checkForMIAs: 56970 events created, 3692 requests dispatched, containing 56368 flattened queries
D/checkForMIAs: 63157 events created, 4090 requests dispatched, containing 62234 flattened queries
D/checkForMIAs: 69110 events created, 4483 requests dispatched, containing 68397 flattened queries
D/checkForMIAs: 75253 events created, 4886 requests dispatched, containing 74400 flattened queries
D/checkForMIAs: 81301 events created, 5278 requests dispatched, containing 80486 flattened queries
D/checkForMIAs: 87529 events created, 5681 requests dispatched, containing 86590 flattened queries
D/checkForMIAs: 93659 events created, 6074 requests dispatched, containing 92723 flattened queries
D/checkForMIAs: 99199 events created, 6450 requests dispatched, containing 98519 flattened queries
D/checkForMIAs: 100000 events created, 6618 requests dispatched, containing 99991 flattened queries
D/checkForMIAs: 100000 events created, 6618 requests dispatched, containing 100000 flattened queries
D/checkForMIAs: All send queries are accounted for.
````
example

### CODE REVIEW

**Comment by d4rken** [comment]
I noticed Piwik does accept http, https and ftp, so I changed the restriction to let through ftp protocol based URLs too. Added a wrapper function for tracking outlinks *trackOutlink(URL)* + tests.

File: piwik_sdk/src/main/java/org/piwik/sdk/Dispatcher.java [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    private Runnable mLoop = new Runnable() {
+        @Override
+        public void run() {
+            android.os.Process.setThreadPriority(Process.THREAD_PRIORITY_BACKGROUND);
+            while (mRunning) {
+                try {
+                    // Either we wait the interval or forceDispatch() granted us one free pass
+                    mSleepToken.tryAcquire(mDispatchInterval, TimeUnit.MILLISECONDS);
```
Comment (dotsbb) [inline piwik_sdk/src/main/java/org/piwik/sdk/Dispatcher.java]: Very clever :+1:

File: piwik_sdk/src/main/java/org/piwik/sdk/Dispatcher.java [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    private boolean doRequest(HttpRequestBase requestBase) {
+        HttpClient client = new DefaultHttpClient();
+        HttpConnectionParams.setConnectionTimeout(client.getParams(), mTimeOut);
+        HttpResponse response;
+
+        if (mPiwik.isDryRun()) {
+            Logy.d(LOGGER_TAG, "DryRun, stored HttpRequest, now " + mDryRunOutput.size());
+            mDryRunOutput.add(requestBase);
+        } else {
+            mDryRunOutput.clear();
```
Comment (dotsbb) [inline piwik_sdk/src/main/java/org/piwik/sdk/Dispatcher.java]: I'll rather check if `mDryRUnOutput` is not empty before trying clearing it (mainly bc of [this](http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/util/ArrayList.java#473))

Comment (d4rken) [inline piwik_sdk/src/main/java/org/piwik/sdk/Dispatcher.java]: It would be a bit more efficient, but I don't think it's noticeable. This happens once per dispatch intervall or once per batched page, i.e. with 2-3digit miliseconds wait times.

I still agree though and will submit another PR for this, but I'd prefer to have this PR merged first instead of further extending this one though.

Or do you mean something different than the efficiency?

File: piwik_sdk/src/main/java/org/piwik/sdk/TrackMe.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+     * set(QueryParams.HOURS, "10");
+     * set(QueryParams.MINUTES, "45");
+     * set(QueryParams.SECONDS, "30");
+     * </pre>
+     *
+     * @param key   query params name
+     * @param value value
+     * @return tracker instance
+     */
+    public synchronized TrackMe set(@NonNull QueryParams key, String value) {
```
Comment (dotsbb) [inline piwik_sdk/src/main/java/org/piwik/sdk/TrackMe.java]: There is any reason why all `TrackMe` public methods are synchronized?

**Comment by dotsbb** [comment]
Thank You for such a great contribution!

File: piwik_sdk/src/main/java/org/piwik/sdk/TrackMe.java (hunk 3 above)
Comment (d4rken) [inline piwik_sdk/src/main/java/org/piwik/sdk/TrackMe.java]: It's an improvement, but not a solution.

Specifically the case of a developer handeling the TrackMe object between two threads, and the developer modifying the TrackMe object after passing it to the Tracker are problems.

Synchronizing the methods against the class was my idea to add some protection against that.
* While doing trySet it will never overwrite a value.
* While doing build the custom variables can not change.
* HashMap is not thread-safe
* CustomVariables are not threadsafe, we don't want them changing when putting them into the params in build()

But I'm not sure now how good this solution is though now you mentioned it.... The developer should not get ideas to try stuff like that anyways ;). If those methods can throw ConcurrencyModificationExceptions, they now can't.
We could go for a ConcurrentHashMap and less locking on the methods as potential improvement, while also making CustomVariables thread-safe.
Performance wise the synchronize statements can be neglected as TrackMe should not be accessed by multiple threads, but i agree that it's not a very clean approach.

**Comment by d4rken** [comment]
@dotsbb 
It wasn't completely altruistic, I needed thread-safety to track events that come from multiple threads in my app :wink: . It's currently in public beta test. Looking good so far.
I also had a little bit fun implementing this :smile:.

Also thank you for all your work! I wouldn't have had the motiviation to do this from scratch. You are not affiliated with Piwik, are you?

Next target would be #16 , don't have a good idea yet though.

**Comment by dotsbb** [comment]
@d4rken I'm a python developer on a daily basis, and don't have a deep knowledge of Piwik at all. Let's say this SDK is side effect of 10 days "hackathon" where I was only participant ;)
